File: 146_LRU_Cache/main.py
class Node:
    def __init__(self, key=None, value=None):
        self.pre = None
        self.next = None
        self.key = key
        self.value = value


class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.his_size = 0

        # history double link-list
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.pre = self.head

    def get(self, key: int) -> int:
        node = self.find(key)
        if node:
            return node.value
        return -1

    def put(self, key: int, value: int) -> None:
        node = self.find(key)
        if node:
            node.value = value
            return

        node = Node(key, value)
        self.cache[key] = node
        self.add_tail(node)

        self.his_size += 1
        if self.his_size > self.capacity:
            to_del = self.head.next
            self.head.next = to_del.next
            to_del.next.pre = self.head
            del self.cache[to_del.key]
            self.his_size = self.capacity

    def find(self, key: int) -> Node:
        if key in self.cache:
            node = self.cache[key]
            if self.tail.pre is not node:
                node.pre.next = node.next
                node.next.pre = node.pre
                self.add_tail(node)
            return node
        return None

    def add_tail(self, node: Node) -> None:
        last = self.tail.pre
        last.next = node
        node.pre = last
        self.tail.pre = node


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)


if __name__ == '__main__':
    # cmd = ["LRUCache", "put", "put", "get", "put", "get", "put", "get", "get", "get"]
    # para = [[2], [1, 1], [2, 2], [1], [3, 3], [2], [4, 4], [1], [3], [4]]
    res = []
    # """
    cache = LRUCache(2)
    res.append(cache.put(1, 1))
    res.append(cache.put(2, 2))
    res.append(cache.get(1))
    res.append(cache.put(3, 3))
    res.append(cache.get(2))
    res.append(cache.put(4, 4))
    res.append(cache.get(1))
    res.append(cache.get(3))
    res.append(cache.get(4))
    """
    cache = LRUCache(1)
    res.append(cache.put(1, 1))
    res.append(cache.put(2, 2))
    res.append(cache.get(1))
    res.append(cache.get(2))
    # """

    for val in res:
        print(val)

File: 146_LRU_Cache/main.cpp
#include <unordered_map>
#include <memory>
using namespace std;

struct Node {
    Node(int key, int value) : key(key), value(value) {}

    shared_ptr<Node> pre = nullptr;
    shared_ptr<Node> next = nullptr;
    int key;
    int value;
};


class LRUCache {
public:
    LRUCache(int capacity) 
        : capacity(capacity)
        , head(make_shared<Node>(-1, -1))
        , tail(make_shared<Node>(-1, -1))
    {
        head->next = tail;
        tail->pre = head;
    }
    
    int get(int key) {
        // if found in cache, move history to most recently, return value
        auto it = cache.find(key);
        if (it != cache.end()) {
            move_to_tail(it->second);
            return it->second->value;
        }
        // if not found in cache, return -1
        else {
            return -1;
        }
    }
    
    void put(int key, int value) {
        // if found in cache, update value and move history to most recently
        auto it = cache.find(key);
        if (it != cache.end()) {
            it->second->value = value;
            move_to_tail(it->second);
        }
        // if not found in cache, set value to move history to most recently
        else {
            // if cache is full, remove least recently used data from cache and history
            auto res = cache.emplace(key, make_shared<Node>(key, value));
            move_to_tail(res.first->second);

            if (cache.size() > capacity) {
                cache.erase(head->next->key);
                auto to_del = head->next;
                head->next = to_del->next;
                head->next->pre = head;
            }
        }
    }

private:
    void move_to_tail(shared_ptr<Node>& node) {
        if (node->next == tail) {
            return;
        }

        // if the node has pre & next, connect them
        if (node->pre && node->next) {
            node->pre->next = node->next;
            node->next->pre = node->pre;
        }
        
        // put the node to tail
        node->pre = tail->pre;
        node->next = tail;
        tail->pre->next = node;
        tail->pre = node;
    }

private:
    int capacity;

    // map[key] = Node
    unordered_map<int, shared_ptr<Node>> cache;

    // double link list of Node for record access history 
    //   (least recently used to most recently used)
    shared_ptr<Node> head;
    shared_ptr<Node> tail;
};

// V test: get non existed node

// V test: put a node when cache is empty
// V test: add new node to tail

// V test: put a node when cache is not empty (found/not_found)
// V test: move the only one node to tail
// V test: move one of multiple nodes to tail

// V test: get a existed node
// V test: phase out a node